includes/admin/plugins.php

```diff
@@ -150,6 +151,126 @@ function perflab_render_plugins_ui() {
 	<?php
 }
 
+/**
+ * Checks if a given plugin is available.
+ *
+ * @since n.e.x.t
+ * @see perflab_install_and_activate_plugin()
+ *
+ * @param array{name: string, slug: string, short_description: string, requires_php: string|false, requires: string|false, requires_plugins: string[], version: string} $plugin_data Plugin data from the WordPress.org API.
+ * @return array{compatible_php: bool, compatible_wp: bool, can_install: bool, can_activate: bool} Availability.
+ */
+function perflab_get_plugin_availability( array $plugin_data ): array {
+	$availability = array(
+		'compatible_php' => (
+			! $plugin_data['requires_php'] ||
+			is_php_version_compatible( $plugin_data['requires_php'] )
+		),
+		'compatible_wp'  => (
+			! $plugin_data['requires'] ||
+			is_wp_version_compatible( $plugin_data['requires'] )
+		),
+	);
+
+	$plugin_status = install_plugin_install_status( $plugin_data );
+
+	// The plugin is already installed or the user can install plugins.
+	$availability['can_install'] = (
+		'install' !== $plugin_status['status'] ||
+		current_user_can( 'install_plugins' )
+	);
+
+	// The plugin is installed and the user can activate it, or the user can activate plugins in general.
+	$availability['can_activate'] = (
+		( $plugin_status['file'] && current_user_can( 'activate_plugin', $plugin_status['file'] ) ) ||
+		current_user_can( 'activate_plugins' )
+	);
+
+	foreach ( $plugin_data['requires_plugins'] as $requires_plugin ) {
+		$dependency_plugin_data = perflab_query_plugin_info( $requires_plugin );
+		if ( $dependency_plugin_data instanceof WP_Error ) {
+			continue;
+		}
+
+		$dependency_availability = perflab_get_plugin_availability( $dependency_plugin_data );
+		foreach ( array( 'compatible_php', 'compatible_wp', 'can_install', 'can_activate' ) as $key ) {
+			$availability[ $key ] = $availability[ $key ] && $dependency_availability[ $key ];
+		}
+	}
+
+	return $availability;
+}
+
+/**
+ * Installs and activates a plugin by its slug.
+ *
+ * Dependencies are recursively installed and activated as well.
+ *
+ * @since n.e.x.t
+ * @see perflab_get_plugin_availability()
+ *
+ * @param string $plugin_slug Plugin slug.
+ * @return WP_Error|null WP_Error on failure.
+ */
+function perflab_install_and_activate_plugin( string $plugin_slug ): ?WP_Error {
```

Comment:
Should we add protection against infinite recursion here by ensuring we only handle a given plugin slug once at most. In theory you if you had a plugin dependency chain that went back to the top plugin, this function would never exit (A->B->C->A). Track which slug has been handled and return early if a handled slug is called.

Reply:
Good question. I was thinking core didn't account for this, but I stand corrected:

> [![image](https://github.com/WordPress/performance/assets/134745/4ebbe51c-f986-4429-98ec-5d05c6859db7)](https://make.wordpress.org/core/2024/03/05/introducing-plugin-dependencies-in-wordpress-6-5/#:~:text=What%20happens%20if%20there%20are%20circular%20dependencies%3F)

Since our plugin has to support 6.4 still we can't reuse the validity-checking logic that core has, but it is overkill for our purposes here since we are managing all of the plugins that we list here. For that matter, checking for recursion may also be overkill since again we're managing all of these plugins, so we hopefully won't make such a recursive error when we list out the `Requires Plugins` field 😄

Reply:
Right, it was mostly a reflexive comment as a general rule for recursive functions - make sure they don't infinitely recurse.

I agree since we control all the plugins here we are unlikely to introduce a circular dependency. Still the code change would be relatively small so maybe worth including anyway? I don't feel strongly so I'll leave it up to you.

Reply:
not sure how core does it, maybe something like https://github.com/WordPress/performance/pull/1197/files

Reply:
@adamsilverstein I had a great conversation with Gemini about this: https://g.co/gemini/share/100242fb5096

I asked about how to address the circular dependency problem and it suggested a dependency tracker as an argument to the function, passed by reference. After some back and forth, I've come up with this: 1362cef.

Reply:
>  had a great conversation with Gemini about this: g.co/gemini/share/100242fb5096

haha, that conversation is quite funny, thanks for sharing. The tips and feedback provided seem pretty through and insightful, I especially like the table with pros and cons of various approaches. Then at then end where it turns out it got some stuff wrong and you admonish it, pure gold.  

>  I've come up with this: 1362cef.

Thats perfect, thanks. Passing the processes list through as a parameter by reference is smart.